File: backend/src/models/Reservation.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const Reservation = sequelize.define('Reservation', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true,
  },
  guestId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: { model: 'guests', key: 'id' },
  },
  roomId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: { model: 'rooms', key: 'id' },
  },
  checkIn: {
    type: DataTypes.DATEONLY,
    allowNull: false,
  },
  checkOut: {
    type: DataTypes.DATEONLY,
    allowNull: false,
  },
  status: {
    type: DataTypes.ENUM('pending', 'confirmed', 'checked_in', 'checked_out', 'cancelled', 'no_show'),
    defaultValue: 'pending',
  },
  adults: {
    type: DataTypes.INTEGER,
    defaultValue: 1,
  },
  children: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
  },
  totalAmount: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: false,
  },
  paidAmount: {
    type: DataTypes.DECIMAL(10, 2),
    defaultValue: 0,
  },
  specialRequests: {
    type: DataTypes.TEXT,
  },
  source: {
    type: DataTypes.ENUM('walk_in', 'phone', 'website', 'booking_com', 'airbnb', 'other'),
    defaultValue: 'walk_in',
  },
  propertyId: {
    type: DataTypes.INTEGER,
    allowNull: true,
    references: { model: 'properties', key: 'id' },
  },
}, {
  tableName: 'reservations',
  timestamps: true,
  paranoid: true,
  indexes: [
    { fields: ['guestId'] },
    { fields: ['roomId'] },
    { fields: ['status'] },
    { fields: ['checkIn'] },
    { fields: ['checkOut'] },
    { fields: ['roomId', 'checkIn', 'checkOut'] },
  ],
});

module.exports = Reservation;

File: mobile/lib/models/reservation.dart
import 'guest.dart';
import 'room.dart';

class Reservation {
  final int id;
  final int guestId;
  final int roomId;
  final DateTime? checkIn;
  final DateTime? checkOut;
  final String status;
  final int adults;
  final int children;
  final double totalAmount;
  final double paidAmount;
  final String? specialRequests;
  final String? source;
  final Guest? guest;
  final Room? room;

  const Reservation({
    required this.id,
    required this.guestId,
    required this.roomId,
    this.checkIn,
    this.checkOut,
    required this.status,
    this.adults = 1,
    this.children = 0,
    required this.totalAmount,
    this.paidAmount = 0,
    this.specialRequests,
    this.source,
    this.guest,
    this.room,
  });

  factory Reservation.fromJson(Map<String, dynamic> json) => Reservation(
        id: _toInt(json['id']) ?? 0,
        guestId: _toInt(json['guestId']) ?? 0,
        roomId: _toInt(json['roomId']) ?? 0,
        checkIn: json['checkIn'] != null ? DateTime.tryParse(json['checkIn'].toString()) : null,
        checkOut: json['checkOut'] != null ? DateTime.tryParse(json['checkOut'].toString()) : null,
        status: json['status']?.toString() ?? 'pending',
        adults: _toInt(json['adults']) ?? 1,
        children: _toInt(json['children']) ?? 0,
        totalAmount: _toDouble(json['totalAmount']) ?? 0,
        paidAmount: _toDouble(json['paidAmount']) ?? 0,
        specialRequests: json['specialRequests']?.toString(),
        source: json['source']?.toString(),
        guest: json['guest'] != null ? Guest.fromJson(json['guest']) : null,
        room: json['room'] != null ? Room.fromJson(json['room']) : null,
      );

  static int? _toInt(dynamic v) => v is int ? v : v is num ? v.toInt() : v is String ? int.tryParse(v) : null;
  static double? _toDouble(dynamic v) => v is double ? v : v is num ? v.toDouble() : v is String ? double.tryParse(v) : null;

  Map<String, dynamic> toJson() => {
        'guestId': guestId,
        'roomId': roomId,
        'checkIn': checkIn?.toIso8601String(),
        'checkOut': checkOut?.toIso8601String(),
        'status': status,
        'adults': adults,
        'children': children,
        'totalAmount': totalAmount,
        'paidAmount': paidAmount,
        'specialRequests': specialRequests,
        'source': source,
      };

  static const statuses = [
    'pending', 'confirmed', 'checked_in', 'checked_out', 'cancelled', 'no_show'
  ];
  static const sources = [
    'walk_in', 'phone', 'website', 'booking_com', 'airbnb', 'other'
  ];
}
